File: code/online_challenges/src/rosalind/algorithmic_heights/fibonacci.py
# Find the nth Fibonacci number
def fib(n):

    fibDict = dict()

    def f(n):

        if n in fibDict:
            return fibDict[n]
        else:
            if n == 0:
                fibDict[0] = 0
                return 0
            elif n == 1:
                fibDict[1] = 1
                return 1
            else:
                fibDict[n] = f(n-1) + f(n-2)
                return fibDict[n]

    return f(n)

File: cosmos-master/code/mathematical-algorithms/fibonacci_number/fibonacci.scala
// 0 and 1 are the first two numbers in the sequence,
//so we start the accumulators with 0,1.
// At every iteration, we add the two numbers to get the next one.
// Path of Cosmos by OpenGenus Foundation
//returns nth fibonacci number
def fib(n: Int): Int = {
@annotation.tailrec//makes sure loop() is tail recursive
 def loop(n: Int, prev: Int, cur: Int): Int =
  if (n == 0) prev
  else loop(n - 1, cur, prev + cur)
loop(n, 0, 1)
}
//to test
//println(fib(6))

